addons/mail/models/mail_message.py

```diff
@@ -391,13 +391,14 @@ def filter_notification(notif):
             for notification in message.notification_ids.filtered(filter_notification):
                 customer_email_data.append((partner_tree[notification.res_partner_id.id][0], partner_tree[notification.res_partner_id.id][1], notification.email_status))
 
-            has_access_to_model = message.model and self.env[message.model].check_access_rights('read', raise_exception=False)
-            main_attachment = has_access_to_model and message.res_id and self.env[message.model].search([('id', '=',message.res_id)]) and getattr(self.env[message.model].browse(message.res_id), 'message_main_attachment_id')
             attachment_ids = []
-            for attachment in message.attachment_ids:
-                if attachment.id in attachments_tree:
-                    attachments_tree[attachment.id]['is_main'] = main_attachment == attachment
-                    attachment_ids.append(attachments_tree[attachment.id])
+            if message.attachment_ids:
```

Comment:
I made the same kind of optimization in https://github.com/odoo/odoo/commit/de7c119e666d108564b3889d1c08e5bf2e4da082#diff-86c5f1ed15bfa23d6be2285dc88e8a57R396
The main difference here is that you are checking message.attachment_ids before computing has_access_to_model.
This is a good idea, but I think that in practice message.model and access rights are in cache. But ok for me.

I just noticed that there is something that is broken in my version since if we don't have access to record but message has attachment, main_attachment will be undefined and code may crash. 

If we want to try to keep the same code between version, i would use issubclass(self.pool[message.model], self.pool['mail.thread']) instead of getattr if possible since I made this change in 12.4 refactoring.